eclass/dub.eclass

```diff
@@ -0,0 +1,269 @@
+DUB_DEPEND=">=dev-util/dub-1.38.0 app-arch/unzip"
+# 1.38 - respect skipRegistry value in configuration file
+
+EDUB_HOME="${WORKDIR}/dub_home"
+
+# @ECLASS_VARIABLE: DUB_OPTIONAL
+# @DEFAULT_UNSET
+# @PRE_INHERIT
+# @DESCRIPTION:
+# Setting this to an non-empty value inhibits the eclass from adding
+# DUB_DEPEND to BDEPEND an exporting phase functions.
+#
+# If set, you may need to call dub_gen_settings or dub_src_unpack
+# manually.
+
+# @ECLASS_VARIABLE: DUB_DEPENDENCIES
+# @DEFAULT_UNSET
+# @PRE_INHERIT
+# @DESCRIPTION:
+# A bash array of dub dependencies.  Each dependency should be a string
+# in the form `name@version`.
+#
+# Example:
+# @CODE
+# DUB_DEPENDENCIES=(
+#     "dfmt@0.15.1"
+#     "silly@1.1.1"
+# )
+# @CODE
+
+# @ECLASS_VARIABLE: DUB_DEPENDENCIES_URIS
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# The transformed DUB_DEPENDENCIES as a string to be placed in
+# SRC_URI
+
+# @FUNCTION: dub_dependencies_uris
+# @USAGE: <dependencies>...
+# @DESCRIPTION:
+# print a string to be placed in SRC_URI for all the dependencies given
+# as arguments.  Each dependency should be in the form `name@version`,
+# just like the elements in DUB_DEPENDENCIES.
+dub_dependencies_uris() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	local registry="https://code.dlang.org/packages"
+	local dep name ver
+	for dep; do
+		name="${dep%@*}"
+		ver="${dep#*@}"
+		echo "${registry}/${name}/${ver}.zip -> ${name}-${ver}.zip"
+	done
+}
+
+# @FUNCTION: _handle_dub_variables
+# @INTERNAL
+# @DESCRIPTION:
+# Verify that DUB_DEPENDENCIES is set appropriately and transform it
+# into DUB_DEPENDENCIES_URIS
+_handle_dub_variables() {
+	[[ -v DUB_DEPENDENCIES && ${DUB_DEPENDENCIES@a} != *a* ]] \
+		&& die "DUB_DEPENDENCIES must be an array"
+
+	[[ ${DUB_OPTIONAL} ]] && return
+
+	BDEPEND+=" ${DUB_DEPEND}"
+
+	[[ ${DUB_DEPENDENCIES[@]} ]] \
+		 && export DUB_DEPENDENCIES_URIS="$(dub_dependencies_uris "${DUB_DEPENDENCIES[@]}")"
+}
+_handle_dub_variables
+unset -f _handle_dub_variables
+
+# @FUNCTION: edub
+# @USAGE: [dub_args]...
+# @DESCRIPTION:
+# A dub wrapper that will automatically set relevant variables and die
+# on failure.
+#
+# Meaningful variables for this function are:
+#
+# $DC - typically set by dlang-r1.eclass or dlang-single.eclass to point
+# to the D compiler and respected by dub.
+#
+# $DFLAGS - dub will use these flags when compiling + linking.  If the
+# variable is unset (empty does not count) this eclass will default to
+# "${DCFLAGS} ${DLANG_LDFLAGS} /-Wno-error/" for its value.
+#
+# $NO_COLOR - if non-empty dub will not print colors
+edub() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	local -x DFLAGS=${DFLAGS-${DCFLAGS} ${DLANG_LDFLAGS} $(dlang_get_wno_error_flag)}
+	local -x DC=${DC} NO_COLOR=${NO_COLOR}
+	# FIXME dub directly calls pkg-config without a way to configure it
+
+	debug-print "${FUNCNAME}: DC=${DC}"
+	debug-print "${FUNCNAME}: DFLAGS=${DFLAGS}"
+
+	set -- dub --verbose "${@}"
+	echo "${@}"
```

Comment:
This is `edo`

Reply:
Bah, how come I only learn of this. I've read the devguide only up to the eclass reference before writing the eclass, such bad timing.